### ISSUE TITLE
Better ergonomics when working with exact size

### ISSUE THREAD

**Issue by kali** [ia0/data-encoding#4]
The following code assert()s unexpectadly. The solution so far is to strip the padding and use the decode_nopad_mut variant.

```rust
        #[test]
        fn test_b64_raw() {
            let a = [0u8; 8];
            let a64 = ::data_encoding::base64::encode(&a);
            assert_eq!("AAAAAAAAAAA=", a64);
            let mut b = [0u8; 8];
            ::data_encoding::base64::decode_mut(a64.as_bytes(), &mut b);
            assert_eq!(a, b);
        }
```

**Comment by ia0** [comment]
Thanks for the issue! This is definitely a shortcoming of `decode_mut`. However, since fixing it would be a breaking change (the documentation specifies that the function should panic in this case), I'll delay this issue to the next major release, which will actually be focused on improving padding issues and allowing callers to decide if and how padding should be handled.

**Comment by ia0** [comment]
In the meantime, the following function finds out automatically the prefix of the input that is necessary to read to fill the output completely:

    fn base64_decode_exact_mut(input: &[u8], output: &mut [u8])
           -> Result<(), data_encoding::decode::Error> {
        let ilen = data_encoding::base64::encode_nopad_len(output.len());
        data_encoding::base64::decode_nopad_mut(&input[0..ilen], output)
    }

**Comment by ia0** [comment]
@kali The [v2.0.0-rc.1][1] now makes the distinction between padded and non-padded bases. You can use `BASE64.no_pad()` to encode and decode non-padded base64.

The v2.0.0-rc.1 should move to v2.0.0 in about one month according to feedback.

[1]: http://ia0.github.io/data-encoding/v2.0.0-rc.1/data_encoding/

**Comment by ia0** [comment]
@kali I don't know if this is still of interest, but I released a v2.0.0-rc.2 and the non-padded encoding is now called `BASE64_NOPAD`.

I'm closing the issue since it's pretty old. Feel free to reopen or create a new issue if your needs are not fulfilled with v2 (should be out of candidate before the end of the year).
